### DISCUSSION TITLE
Hillshading incompatible with Android SAF

### DISCUSSION THREAD

**Discussion by ztNFny** [mapsforge/mapsforge#1269] [Ideas]
While adding hillshading to an Android app I ran into the issue that all the hillshading classes currently expect a `java.io.File` object for the folder containing the hgt/dem files. Android 11 introduced Storage Access Framework (SAF) which block general access to the file system and only grants access to specific folders the user must explicitly allow. However even for those folders you don't get a File object (but can read them using an InputStream).

The mapsforge code handling the file operations is in the common part (mapsforge-map/src/main/java/org/mapsforge/map/layer/hills/), so "just" changing that code isn't possible as it must continue working for all other platforms but an abstraction layer will be required.

There are two main classes which are using doing `File` operations:
- https://github.com/mapsforge/mapsforge/blob/e558dea5e9a22d5c9b500d01c2a6b9b92b100643/mapsforge-map/src/main/java/org/mapsforge/map/layer/hills/HgtCache.java#L159 crawls the filesystem and builds an index of files
- https://github.com/mapsforge/mapsforge/blob/e558dea5e9a22d5c9b500d01c2a6b9b92b100643/mapsforge-map/src/main/java/org/mapsforge/map/layer/hills/AbsShadingAlgorithmDefaults.java#L43 does the actual file reading

As this code is in mapsforge core and a change for SAF will require at least one abstraction layer this change would certainly be all but trivial - so I'm posting here before even starting to look at it further. Is this something you're interested in supporting or maybe even something already in work?  Would you support such a change?

**Comment by devemux86** [comment]
Hillshading was contributed by the community (@usrusr).

Someone needs to modify it to work **also** with file streams,
in addition to how now works with regular files.

So it can work on all platforms, like the map files and render themes.

**Comment by devemux86** [comment]
See https://github.com/mapsforge/mapsforge/pull/1338.
